### ISSUE TITLE
Topological sort - bug

### ISSUE THREAD

**Issue by gordicaleksa** [karpathy/micrograd#67]
It's a nit that won't matter most of the time but the topo sort [implementation](https://github.com/karpathy/micrograd/blob/master/micrograd/engine.py#L59) doesn't work in case you have cycles in the graph.

i.e. there is a hard assumption you're operating over a DAG.

**Comment by strisys** [comment]
Why can't this function simply be implemented as follows?

```
  def backward(self, is_first=True):
    if (is_first == True):
      self.grad = 1.0

    self._backward()

    for c in self._prev:
      c.backward(False)
```

**Comment by Yara97Mansour** [comment]
Do you mean when there are two or more current children nodes are connected and you want to conceder one of them as a parent node to the other children ones?

If that right you can just do the following:

def backward(self):

        # topological order all of the children in the graph
        topo = []
        visited = dict()
        def build_topo(v):
            if v not in visited.keys():
                visited[v] = v._prev
                for child in visited[v]:
                    build_topo(child)
                topo.append(v)
        build_topo(self)

NOTE: I did not test it yet, but I think using dictionary instead of "set" to define the children of each node separately should be fine!

**Comment by Narasimhareddy-B** [comment]
A directed acyclic graph is a directed graph that has no cycles. A vertex v of a directed graph is said to be reachable from another vertex u when there exists a path that starts at u and ends at v. As a special case, every vertex is considered to be reachable from itself (by a path with zero edges).

**Comment by Jet-lag** [comment]
> It's a nit that won't matter most of the time but the topo sort [implementation](https://github.com/karpathy/micrograd/blob/master/micrograd/engine.py#L59) doesn't work in case you have cycles in the graph.
> 
> i.e. there is a hard assumption you're operating over a DAG.

It's correct. On the other hand, if it is a Dag, can we simply write the backward function as 
```
def backward(self):  
    self._backward()  
    for child in self._prev:  
      child.backward()
```

**Comment by pkulijing** [comment]
> > It's a nit that won't matter most of the time but the topo sort [implementation](https://github.com/karpathy/micrograd/blob/master/micrograd/engine.py#L59) doesn't work in case you have cycles in the graph.
> > i.e. there is a hard assumption you're operating over a DAG.
> 
> It's correct. On the other hand, if it is a Dag, can we simply write the backward function as
> 
> ```
> def backward(self):  
>     self._backward()  
>     for child in self._prev:  
>       child.backward()
> ```

No. Your implementation is essentially a DFS, while topological sort requires a BFS. For this simple case your code could be wrong:
```
b = 2*a
c = a + b
```
with topological sort, it's guaranteed that the back propagation goes in the order of c -> b -> a. With your code, the order could be c -> a -> b, which is wrong. 
```

**Comment by conscell** [comment]
@gordicaleksa
Could you please provide an example when the computational graph is not a DAG?

**Comment by pc-cp** [comment]
> > > It's a nit that won't matter most of the time but the topo sort [implementation](https://github.com/karpathy/micrograd/blob/master/micrograd/engine.py#L59) doesn't work in case you have cycles in the graph.
> > > i.e. there is a hard assumption you're operating over a DAG.
> > 
> > 
> > It's correct. On the other hand, if it is a Dag, can we simply write the backward function as
> > ```
> > def backward(self):  
> >     self._backward()  
> >     for child in self._prev:  
> >       child.backward()
> > ```
> 
> No. Your implementation is essentially a DFS, while topological sort requires a BFS. For this simple case your code could be wrong:
> 
> ```
> b = 2*a
> c = a + b
> ```
> 
> with topological sort, it's guaranteed that the back propagation goes in the order of c -> b -> a. With your code, the order could be c -> a -> b, which is wrong.

l try use DFS, find something interesting: l print node traversal order
first, use topo sort:
 ```
def backward(self):
    # topological order all of the children in the graph
    topo = []
    visited = set()
    def build_topo(v):
        if v not in visited:
            visited.add(v)
            for child in v._prev:
                build_topo(child)
            topo.append(v)
    build_topo(self)

    # go one variable at a time and apply the chain rule to get its gradient
    self.grad = 1.0
    for node in reversed(topo):
        node._backward()
    print([i for i in reversed(topo)])

    # print(self)
    # self._backward()  
    # for child in self._prev:  
    #   child.backward()
```

<img width="1176" height="272" alt="Image" src="https://github.com/user-attachments/assets/7cba63af-755f-4591-85e6-59e8539de044" />

second, use DFS way:
```
def backward(self):
    # topological order all of the children in the graph
    # topo = []
    # visited = set()
    # def build_topo(v):
    #     if v not in visited:
    #         visited.add(v)
    #         for child in v._prev:
    #             build_topo(child)
    #         topo.append(v)
    # build_topo(self)

    # # go one variable at a time and apply the chain rule to get its gradient
    # self.grad = 1.0
    # for node in reversed(topo):
    #     node._backward()
    # print([i for i in reversed(topo)])

    print(self)
    self._backward()  
    for child in self._prev:  
      child.backward()
```
it print:

<img width="1196" height="346" alt="Image" src="https://github.com/user-attachments/assets/231476c9-b2db-420d-a9dc-4afcf57e2241" />

although both way print same grad value, but l have to say: node `a` calculate twice. first from `b`, second directly from `c`. if this way use in neural network, it will lead `floating-point error, the deeper the network leads to more severe error accumulation.`
